internal/backend/remote-state/s3/backend.go

```diff
@@ -1171,19 +1178,28 @@ func (e customEndpoint) String(obj cty.Value) string {
 	return v
 }
 
+func includeProtoIfNessesary(endpoint string) string {
+	//if !strings.HasPrefix(endpoint, "http://") || strings.HasPrefix(endpoint, "https://") {
+	if matched, _ := regexp.MatchString("[a-z]*://.*", endpoint); !matched {
+		log.Printf("[DEBUG] Adding https:// prefix to endpoint '%s'", endpoint)
+		endpoint = fmt.Sprintf("https://%s", endpoint)
+	}
+	return endpoint
+}
```

Comment:
I think instead of using regex, I would prefer if we parse the url and check the scheme directly.

Something like this should work I think

```suggestion
func includeProtoIfNecessary(endpoint string) (string, error) {
	parsedUrl, err := url.Parse(endpoint)
	if err != nil {
		return endpoint, err
	}

	if parsedUrl.Scheme == "" {
		log.Printf("[DEBUG] Adding https:// prefix to endpoint '%s'", endpoint)
		endpoint = "https://" + endpoint
	}

	return endpoint
}
```

Reply:
You would have to add the errorhandling higher up though for the failure to parse

Reply:
Unfortunately, go's url.Parse(s) does not handle proto-less urls in a good way.  In many cases, it will set Scheme to the hostname, among other things.  That's why terraform's similar check looks at both Scheme != "" and Host != "".  I think the current approach is cleaner as we are not relying on a edge case in url.Parse.

Reply:
Sounds good to me then! 👍